### ISSUE TITLE
Generated runner scripts use absolute node path

### ISSUE THREAD

**Issue by mgsloan** [ghcjs/ghcjs#491]
This absolute path is written here:  https://github.com/ghcjs/ghcjs/blob/7aa7bae05b306bdb8f7a9c6e292ce74736e4541d/src/Gen2/Linker.hs#L373

For example, I might end up with a js file that has "#!/usr/bin/node".  In my case, I am building in a docker container which does have a node binary in that location.  However, these scripts now do not work on the host machine, because I have node installed in `~/.local/bin`.

Perhaps `/usr/bin/env` should be used?  See http://unix.stackexchange.com/questions/29608/why-is-it-better-to-use-usr-bin-env-name-instead-of-path-to-name-as-my

**Comment by bergey** [comment]
Sounds good.  Can you put together a PR?  I can review.

**Comment by luite** [comment]
The reason for using an absolute path was for the TH runner and the interpreter. `ghcjs-boot` resolves the absolute path and stores it in this special file.

I'm cleaning up the code and moving this to the `settings` file (ghc 8 branch only), where other commands are stored as well, with absolute paths:

```
 ("Haskell CPP command","/usr/bin/gcc"),
 ("Haskell CPP flags","-E -undef -traditional -Wno-invalid-pp-token -Wno-unicode -Wno-trigraphs"),
 ("ld command", "/usr/bin/ld"),
 ("ld flags", " -arch x86_64"),
 ("ld supports compact unwind", "YES"),
 ("ld supports build-id", "NO"),
 ("ld supports filelist", "YES"),
 ("ld is GNU ld", "NO"),
 ("ar command", "/usr/bin/ar"),
```

I think the absolute path should be kept at least for build purposes. Should we have another setting when generating runnable wrapper scripts?

**Comment by spl** [comment]
Is it possible for `ghcjs` to use the `node` in the `$PATH` unless overridden by a setting or flag?

**Comment by luite** [comment]
would that be a good idea? The name of the program doesn't seem consistent: Debian/Ubuntu use `nodejs`. Using the absolute path at least guarantees that Cabal Setup scripts work locally, as long as GHCJS has been set up correctly.

An alternative: Is it possible to write a shebang line that tries in this order:

 - Hardcoded absolute node.js path
- `nodejs` program in `$PATH`
- `node` program in `$PATH`

**Comment by sumitsahrawat** [comment]
@luite: http://unix.stackexchange.com/a/65295/54081

**Comment by spl** [comment]
@luite Right, that's basically what I'm saying. Though I would use this order:

1. Command-line flag path (`--with-node` or whatever you call it)
2. `settings` absolute file path for `nodejs` or `node`
3. `nodejs` in `$PATH`
4. `node` in `$PATH`

I'm not sure what the `settings` file is, though you mention it above.

[`command`](http://pubs.opengroup.org/onlinepubs/009696899/utilities/command.html) is a wonderful POSIX command. I suppose there's also a [Windows alternative](https://stackoverflow.com/questions/304319/is-there-an-equivalent-of-which-on-the-windows-command-line), though I'm not familiar with it.

**Comment by boj** [comment]
I blew 1.5 days troubleshooting why things wouldn't build after swapping my node install from brew to nix on OSX (among other things, so was hard to pin down the cause). Turns out it was this hard coded path 😐
